### PR TITLE
sys-kernel/kpatch: init

### PR DESCRIPTION
**Pull request by aliceinwire**
new kpatch package
the ebuild is based on maffblaster ebuild

### CODE REVIEW

File: sys-kernel/kpatch/kpatch-0.4.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+else
+	SRC_URI="https://github.com/dynup/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+	KEYWORDS="~amd64"
+fi
+
+DESCRIPTION="Dynamic kernel patching for Linux"
+HOMEPAGE="https://github.com/dynup/kpatch"
+LICENSE="GPL-2"
+SLOT="0"
+IUSE="test examples +modules"
```
Comment (mgorny) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: Sort lexically.

File: sys-kernel/kpatch/kpatch-0.4.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+DEPEND="
+	sys-devel/bison
```
Comment (mgorny) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: Sorting ;-).

File: sys-kernel/kpatch/kpatch-0.4.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	dev-libs/elfutils
+	>sys-devel/gcc-4.9.3
```
Comment (mgorny) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: Compiler dependencies in DEPEND are meaningless due to gcc-config. You need to use tc-* functions.

File: sys-kernel/kpatch/kpatch-0.4.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	inherit git-r3
+	EGIT_REPO_URI="https://github.com/dynup/${PN}.git"
+else
+	SRC_URI="https://github.com/dynup/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+	KEYWORDS="~amd64"
+fi
+
+DESCRIPTION="Dynamic kernel patching for Linux"
+HOMEPAGE="https://github.com/dynup/kpatch"
+LICENSE="GPL-2"
```
Comment (SoapGentoo) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: empty line before

File: sys-kernel/kpatch/kpatch-0.4.0.ebuild (hunk 3 above)
Comment (SoapGentoo) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: why do you have this dependency in here? depending on gcc in `DEPEND` makes little sense

File: sys-kernel/kpatch/kpatch-0.4.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+DEPEND="
+	sys-devel/bison
+	dev-libs/elfutils
+	>sys-devel/gcc-4.9.3
+"
+
+RDEPEND="
+	app-crypt/pesign
+	dev-libs/openssl
```
Comment (SoapGentoo) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: does it link to openssl? If so, `:0=`

Comment (mgorny) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: Doesn't it require those libraries at build time as well?

File: sys-kernel/kpatch/kpatch-0.4.0.ebuild [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	sys-libs/zlib
+	sys-apps/pciutils
+	sys-libs/ncurses:5
```
Comment (SoapGentoo) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: same here, are you sure you really want `SLOT=5`? this is a binary-only slot

File: sys-kernel/kpatch/kpatch-0.4.0.ebuild [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+			CONFIG_CHECK="FUNCTION_TRACER HAVE_FENTRY MODULES SYSFS KALLSYMS_ALL"
+			ERROR_FUNCTION_TRACER="CONFIG_FUNCTION_TRACER must be enabled in the kernel's config file"
+			ERROR_HAVE_FENTRY="CONFIG_HAVE_FENTRY must be enabled in the kernel's config file"
+			ERROR_MODULES="CONFIG_MODULES must be enabled in the kernel's config file"
+			ERROR_SYSFS="CONFIG_SYSFS must be enabled in the kernel's config file"
+			ERROR_KALLSYMS_ALL="CONFIG_KALLSYMS_ALL must be enabled in the kernel's config file"
+		fi
+	else
+		eerror "kpatch is not available for Linux kernels below 4.0.0"
+		eerror "Upgrade the kernel sources before installing kpatch." && die
```
Comment (mgorny) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: You can't rely on `eerror` returning success for `&&`.

File: sys-kernel/kpatch/kpatch-0.4.0.ebuild [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	fi
+
+	check_extra_config
+}
+
+src_prepare() {
+	replace-flags -O? -O1
```
Comment (mgorny) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: Quote the pattern to prevent accidental filename expansion.

File: sys-kernel/kpatch/metadata.xml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+<maintainer type="person">
+<email>alicef@gentoo.org</email>
+</maintainer>
+<longdescription lang="en">
+Dynamic kernel patching for the Linux kernel
+</longdescription>
+</pkgmetadata>
```
Comment (SoapGentoo) [inline sys-kernel/kpatch/metadata.xml]: please do some minimal indenting here, preferably using `xmllint`

File: sys-kernel/kpatch/metadata.xml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+<maintainer type="person">
+<email>alicef@gentoo.org</email>
+</maintainer>
+<longdescription lang="en">
+Dynamic kernel patching for the Linux kernel
```
Comment (mgorny) [inline sys-kernel/kpatch/metadata.xml]: Either remove longdescription or put a *long* description there. Repeating DESCRIPTION makes no sense.

File: sys-kernel/kpatch/kpatch-0.4.0.ebuild [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	default
+}
+
+src_install() {
+	unset ARCH
```
Comment (mgorny) [inline sys-kernel/kpatch/kpatch-0.4.0.ebuild]: You shouldn't rely on those vars not being readonly. Try `ARCH=` to make.

**Pull request closed without merging**